File: src/models/purchase.py
from sqlalchemy import ForeignKey, Integer, String
from sqlalchemy.orm import Mapped, mapped_column

from src.database.connection import Base


class Purchase(Base):
    __tablename__ = "purchases"

    id: Mapped[int] = mapped_column(
        primary_key=True,
        autoincrement=True,
    )

    supplier_id: Mapped[int] = mapped_column(
        Integer,
        ForeignKey("suppliers.id"),
        nullable=False,
    )

    invoice_number: Mapped[str] = mapped_column(
        String(100),
        nullable=False,
    )

    invoice_series: Mapped[str | None] = mapped_column(
        String(50),
        nullable=True,
    )

    issue_date: Mapped[str] = mapped_column(
        String(30),
        nullable=False,
    )

    received_at: Mapped[str | None] = mapped_column(
        String(30),
        nullable=True,
    )

    notes: Mapped[str | None] = mapped_column(
        String(1000),
        nullable=True,
    )

    created_by: Mapped[int] = mapped_column(
        Integer,
        ForeignKey("users.id"),
        nullable=False,
    )

    created_at: Mapped[str] = mapped_column(
        String(30),
        nullable=False,
    )

    updated_at: Mapped[str] = mapped_column(
        String(30),
        nullable=False,
    )

    status: Mapped[str] = mapped_column(
        String(50),
        nullable=False,
    )

File: frontend/src/types/purchase.ts
export type PurchaseStatus =
  | 'PENDING'
  | 'RECEIVED'
  | 'CANCELLED'

export interface Purchase {
  id: number
  supplier_id: number

  invoice_number: string
  invoice_series: string | null

  issue_date: string
  received_at: string | null

  notes: string | null

  created_by: number
  created_at: string
  updated_at: string

  status: PurchaseStatus
}

export interface PurchaseItem {
  id: number
  purchase_id: number
  part_id: number

  quantity_purchased: number
  quantity_available: number

  created_at: string
}

export interface PurchaseCreatePayload {
  supplier_id: number
  invoice_number: string
  invoice_series?: string | null
  issue_date: string
  status?: PurchaseStatus
  notes?: string | null
}

export interface PurchaseUpdatePayload {
  supplier_id?: number
  invoice_number?: string
  invoice_series?: string | null
  issue_date?: string
  status?: PurchaseStatus
  notes?: string | null
}

export interface PurchaseCancelPayload {
  justification: string
}

export interface PurchaseItemCreatePayload {
  part_id: number
  quantity_purchased: number
}

export interface PurchaseFormValues {
  supplier_id: number
  invoice_number: string
  invoice_series: string | null
  issue_date: string
  notes: string | null
}